File: examples/aoc/2025/1/main.zig
const std = @import("std");

fn mod(x: i64, y: i64) i64 {
    return @rem(@rem(x, y) + y, y);
}

pub fn main() !void {
    const allocator = std.heap.page_allocator;

    var threaded = std.Io.Threaded.init(allocator, .{});
    defer threaded.deinit();
    const io = threaded.io();

    const file = try std.Io.Dir.cwd().openFile(io, "input.txt", .{});
    defer file.close(io);

    const stat = try file.stat(io);
    const contents = try allocator.alloc(u8, stat.size);
    defer allocator.free(contents);
    _ = try file.readPositionalAll(io, contents, 0);

    var nums: [5000]i64 = undefined;
    var n: usize = 0;

    var i: usize = 0;
    while (i < contents.len) {
        const start = i;
        while (i < contents.len and contents[i] != '\n') : (i += 1) {}
        const end = i;

        if (end == start) {
            nums[n] = 0;
            n += 1;
        } else {
            const sign: i64 = if (contents[start] == 'L') -1 else 1;
            var val: i64 = 0;
            var j: usize = start + 1;
            while (j < end) : (j += 1) {
                val = val * 10 + (contents[j] - '0');
            }
            nums[n] = val * sign;
            n += 1;
        }
        if (i >= contents.len) break;
        i += 1;
    }
    if (contents.len > 0 and contents[contents.len - 1] == '\n') {
        nums[n] = 0;
        n += 1;
    }

    var dial: i64 = 50;
    var count: i64 = 0;
    for (nums[0..n]) |x| {
        dial = mod(dial + x, 100);
        if (dial == 0) count += 1;
    }

    std.debug.print("part_1: {d}\n", .{count});

    dial = 50;
    count = 0;
    for (nums[0..n]) |x| {
        const s = dial + x;
        if (dial > 0 and x < 0 and s < 0) count += 1;
        const abs_s = if (s >= 0) s else -s;
        const adj = @max(abs_s - 1, 0);
        count += @divTrunc(adj, 100);
        dial = mod(dial + x, 100);
        if (dial == 0) count += 1;
    }
    std.debug.print("part_2: {d}\n", .{count});
}

File: examples/aoc/2025/1/main.cpp
#include <cstdint>
#include <cstdio>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>

static int64_t mod(int64_t x, int64_t y) {
    return ((x % y + y) % y);
}

int main() {
    std::ifstream f("input.txt");
    std::string content((std::istreambuf_iterator<char>(f)), std::istreambuf_iterator<char>());

    std::vector<int64_t> nums;
    size_t pos = 0;
    while (pos < content.size()) {
        size_t next = content.find('\n', pos);
        if (next == std::string::npos)
            next = content.size();
        if (pos == next) {
            nums.push_back(0);
        } else {
            int64_t sign = (content[pos] == 'L') ? -1 : 1;
            int64_t val = std::stoll(content.substr(pos + 1, next - pos - 1));
            nums.push_back(val * sign);
        }
        pos = next + 1;
        if (next == content.size())
            break;
    }
    if (!content.empty() && content.back() == '\n')
        nums.push_back(0);

    int64_t dial = 50, count = 0;
    for (auto x : nums) {
        dial = mod(dial + x, 100);
        if (dial == 0)
            count++;
    }
    printf("part_1: %ld\n", (long)count);

    dial = 50;
    count = 0;
    for (auto x : nums) {
        if (dial > 0 && x < 0 && dial + x < 0)
            count++;
        int64_t adj = llabs(dial + x) - 1;
        count += adj / 100;
        dial = mod(dial + x, 100);
        if (dial == 0)
            count++;
    }
    printf("part_2: %ld\n", (long)count);
    return 0;
}
